File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if(lists.length == 0) return null;
        while(lists.length > 1)
        {
            ListNode[] newLists = new ListNode[(lists.length + 1)/2];
            int p = 0;
            for(int i = 0; i < lists.length; i+=2)
            {
                if(i == lists.length - 1)
                    newLists[p++] = lists[i];
                else
                    newLists[p++] = merge(lists[i], lists[i + 1]);
            }
            
            lists = newLists;
        }
        
        return lists[0];
    }
    public ListNode merge(ListNode list1, ListNode list2){
        ListNode pointer1 = list1;
        ListNode pointer2 = list2;
        ListNode dummy = new ListNode(-1);
        ListNode cursor = dummy;

        while (pointer1 != null && pointer2 != null){
            if (pointer1.val < pointer2.val){
                cursor.next = pointer1;
                pointer1 = pointer1.next;
                cursor = cursor.next;
            }
            else{
                cursor.next = pointer2;
                pointer2 = pointer2.next;
                cursor = cursor.next;
            }
        }
        while (pointer1 != null){
            cursor.next = pointer1;
            pointer1 = pointer1.next;
            cursor = cursor.next;
        }
        while (pointer2 != null){
            cursor.next = pointer2;
            pointer2 = pointer2.next;
            cursor = cursor.next;
        }
        return dummy.next;
    }
}

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* merge(ListNode* l1, ListNode* l2)
    {
        ListNode* head = new ListNode(0);
        ListNode* cursor = head;
        
        while(l1 != nullptr && l2 != nullptr)
        {
            // cout << l1 << " " << l2 << endl;
            
            if(l1->val < l2->val)
            {
                cursor->next = l1;
                l1 = l1->next;
            }
            else
            {
                cursor->next = l2;
                l2 = l2->next;
            }
            
            cursor = cursor->next;
        }
        
        while(l1 != nullptr)
        {
            cursor->next = l1;
            l1 = l1->next;
            cursor = cursor->next;
        }
        
        while(l2 != nullptr)
        {
            cursor->next = l2;
            l2 = l2->next;
            cursor = cursor->next;
        }
        
        return head->next;
    }
    
    ListNode* mergeKLists(vector<ListNode*>& lists)
    {
        if(lists.size() == 0) return nullptr;
        
        while(lists.size() > 1)
        {
            vector<ListNode*> cand;
            
            for(int i = 0; i < lists.size() - 1; i+=2)
            {
                cand.push_back(merge(lists[i], lists[i + 1]));
            }
            
            if(lists.size() % 2 == 1)
            {
                cand.push_back(lists[lists.size() - 1]);
            }
            
            lists = cand;
        }
        
        return lists[0];
    }
};